examples/streaming/wordcount/src/main/scala/org/apache/gearpump/streaming/examples/wordcount/dsl/refactor/WordCount.scala

```diff
@@ -0,0 +1,151 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.streaming.examples.wordcount.dsl.refactor
+
+import org.apache.commons.logging.{Log, LogFactory}
+import org.apache.gearpump.streaming.refactor.coder.{StringUtf8Coder, VarLongCoder}
+import org.apache.gearpump.streaming.refactor.dsl.api.functions.MapWithStateFunction
+import org.apache.gearpump.streaming.refactor.dsl.scalaapi.StreamApp
+import org.apache.gearpump.streaming.refactor.dsl.scalaapi.functions.FlatMapWithStateFunction
+import org.apache.gearpump.streaming.refactor.state.api.{StateInternals, ValueState}
+import org.apache.gearpump.streaming.refactor.state.{RuntimeContext, StateNamespaces, StateTags}
+import org.apache.hadoop.conf.Configuration
+import org.apache.gearpump.cluster.UserConfig
+import org.apache.gearpump.cluster.client.ClientContext
+import org.apache.gearpump.cluster.main.{ArgumentsParser, CLIOption, ParseResult}
+import org.apache.gearpump.streaming.hadoop.HadoopCheckpointStoreFactory
+import org.apache.gearpump.streaming.hadoop.lib.rotation.FileSizeRotation
+import org.apache.gearpump.streaming.state.impl.PersistentStateConfig
+import org.apache.gearpump.streaming.{Processor, StreamApplication}
+import org.apache.gearpump.util.{AkkaApp, Graph}
+
+/**
+ *
+ */
+object WordCount extends AkkaApp with ArgumentsParser {
+
+  override val options: Array[(String, CLIOption[Any])] = Array.empty
+
+  override def main(akkaConf: Config, args: Array[String]): Unit = {
+    val context = ClientContext(akkaConf)
+
+    val hadoopConfig = new Configuration
+    val checkpointStoreFactory = new HadoopCheckpointStoreFactory("MessageConsume", hadoopConfig,
+      // Rotates on 1MB
+      new FileSizeRotation(1000000))
+    val taskConfig = UserConfig.empty
+      .withBoolean(PersistentStateConfig.STATE_CHECKPOINT_ENABLE, true)
+      .withLong(PersistentStateConfig.STATE_CHECKPOINT_INTERVAL_MS, 1000L)
+      .withValue(PersistentStateConfig.STATE_CHECKPOINT_STORE_FACTORY,
+        checkpointStoreFactory)(context.system)
+
+    val app = StreamApp("dsl", context, taskConfig)
+    val data = "This is a good start, bingo!! bingo!!"
+    app.source(data.lines.toList, 1, "source").
+      // word => (word, count)
+      flatMapWithState(new StatefulFlatMapFunction(), "a stateful flatmap udf").
+      mapWithState(new StatefulMapFunction(), "").
+      // (word, count1), (word, count2) => (word, count1 + count2)
+      groupByKey().sum.log
+
+    context.submit(app).waitUntilFinish()
+    context.close()
+  }
+
+
+  private class StatefulFlatMapFunction
+    extends FlatMapWithStateFunction[String, String] {
+
+    private val logger: Log = LogFactory.getLog(getClass)
+
+    private implicit val counterStateTag = "tag1"
```

Comment:
We'd better refrain us from using Scala implicit in user land and provide an explicit interface instead